plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/InternalGoogleCloudStorageService.java

```diff
@@ -0,0 +1,330 @@
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import static java.net.HttpURLConnection.HTTP_FORBIDDEN;
+import static java.net.HttpURLConnection.HTTP_NOT_FOUND;
+
+public class InternalGoogleCloudStorageService extends AbstractComponent implements GoogleCloudStorageService {
+
+    /**
+     * Google Cloud Storage batch requests are limited to 1000 operations
+     **/
+    private static final int MAX_BATCHING_REQUESTS = 1_000;
+
+    @Inject
+    public InternalGoogleCloudStorageService(Settings settings) {
+        super(settings);
+    }
+
+    @Override
+    public Client client(Path credentials, String application) {
+        try {
+            logger.debug("loading Google Cloud Storage credentials file from [{}]", credentials);
+            GoogleCredential credential = loadCredentials(credentials);
+
+            logger.debug("building HTTP transport");
+            NetHttpTransport.Builder httpTransport = new NetHttpTransport.Builder();
+
+            logger.debug("creating Google Cloud Storage client for application [{}]", application);
+            Storage.Builder storage = new Storage.Builder(httpTransport.build(), JacksonFactory.getDefaultInstance(), credential);
+            storage.setApplicationName(application);
+
+            return new InternalGcsClient(storage.build());
+        } catch (IOException e) {
+            throw new ElasticsearchException("Error when loading Google Cloud Storage credentials file");
+        }
+    }
+
+    private GoogleCredential loadCredentials(Path credentialsFilePath) throws IOException {
+        if (credentialsFilePath == null) {
+            throw new ElasticsearchException("Cannot load Google Cloud Storage credentials from a null path");
+        }
+        try (InputStream is = Files.newInputStream(credentialsFilePath)) {
+            GoogleCredential credential = GoogleCredential.fromStream(is);
+            if (credential.createScopedRequired()) {
+                credential = credential.createScoped(StorageScopes.all());
+            }
+            return credential;
+        }
+    }
+
+    class InternalGcsClient implements Client {
+
+        private final Storage client;
+
+        public InternalGcsClient(Storage client) {
+            this.client = client;
+        }
+
+        @Override
+        public boolean doesBucketExist(String bucketName) throws IOException {
+            logger.debug("checking if bucket [{}] exists", bucketName);
+            try {
+                Bucket bucket = client.buckets().get(bucketName).execute();
+                if (bucket != null) {
+                    return Strings.hasText(bucket.getId());
+                }
+            } catch (GoogleJsonResponseException e) {
+                if (e.getStatusCode() == HTTP_FORBIDDEN) {
+                    logger.warn("access to bucket [{}] is forbidden", bucketName);
+                    return false;
```

Comment:
I am not sure that false is correct answer here since if the bucket doesn't exist we will try to create it in the next step. Would it make more sense to bail out with an exception right here.